Junk4Dollars-app/API/ApiClient.swift

```diff
@@ -1,5 +1,6 @@
 import Foundation
 
 public protocol ApiClient {
-    func makeApiCall(endpoint: String, whatToDoWithResponseData: @escaping (Data?, URLResponse?, Error?) -> Void)
+    func makePublicApiCall(endpoint: String, callback: @escaping (Data?, URLResponse?, Error?) -> Void)
+    func makeAuthorizedApiCall(endpoint: String, callback: @escaping (Data?, URLResponse?, Error?) -> Void)
```

Comment:
Each individual function has become easier to read on its own, however I am not sure this addresses the underlying issue. The high level goal is to have a stable interface. Two questions:

1. Using the same hypothetical example as before, how would this protocol / interface have to change to support calling an endpoint to create an auction?
2. Presuming this protocol would have to change in order to support that feature, can you think of a refactoring you could do _now_ so that this protocol would be unlikely to change in order to support that feature?

Reply:
I should definitely change the method names to be specific to getting data from my api. I'm not exactly sure what the syntax would look like for uploading data (ex/ making a new auction) but it's likely:
```
    public func sendDataToApi(endpoint: String, data: Any, callback: @escaping (Data?, URLResponse?, Error?) -> Void) {
        AuthenticationDependencies.authentication.getAccessToken() { accessToken in
            let task = self.createURLSessionUploadTask(endpoint: endpoint, accessToken: accessToken, data: Any, callback: callback)
            task.resume()
        }
    }
```

It looks similar to getting data with authorization, but it'd need to accept data to upload and would use a different private method to create an `uploadTask`. I'm not sure what logic could be shared between this and the current GET methods though.

Another question that just came up is how do I type the data for this `sendDataToApi` method? It could be an Auction or a User at this point. I'm not implementing this yet, but just wondering.

Reply:
Let's ignore the implementation for the moment, and just focus on the interface. Can you think of a way to refactor this interface for the current featureset, so that it would not have to change for sending data? Basically, the goal is to make it so that the `ApiClient` protocol never changes.

Reply:
I was thinking you meant abstracting out some logic from the current methods that might be shared between them and an eventual 'send data' method, so that the current methods on the protocol wouldn't need to be renamed or refactored, but do you actually mean how can I make this protocol not change at all - as in the current methods are all it will have? I guess you could have it take more parameters and have it either retrieve or send data depending on whether a `callApi` method is handed data, but that seems like I'm overly generalizing the method and trying to anticipate how I'll need it to work.

Reply:
Gotcha. I did mean so that the protocol would not change at all.

Reply:
So perhaps making a more generic `makeApiCall` method that works for both unauthorized / authorized calls AND downloading / uploading data? That would make it so the protocol wouldn't change. When you say the protocol doesnt change, do you mean no new code or no new publicly accessible methods / variables? I could generalize it more so there's one method you'd interface with, and just add a new private method in the future when we do uploading for new auctions / users.